### PR TITLE
En 7795/esdt pause

### PR DESCRIPTION
**Pull request by sasurobert**
implemented ESDT pause functionality. It will set a flag in all the shards to pause the transactions of a certain token.

### CODE REVIEW

File: vm/systemSmartContracts/esdt.go [hunk 1]
```diff
@@ -407,10 +408,6 @@ func (e *esdt) toggleFreeze(args *vmcommon.ContractCallInput, builtInFunc string
 		e.eei.AddReturnMessage("cannot freeze")
 		return vmcommon.UserError
 	}
-	if len(args.Arguments[1]) != len(args.CallerAddr) {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/esdt.go:410]: if len(args.Arguments) != 2 in line 398 and 426?

File: process/transaction/metaProcess.go [hunk 2]
```diff
@@ -102,6 +118,8 @@ func (txProc *metaTxProcessor) ProcessTransaction(tx *transaction.Transaction) (
 		return txProc.processSCDeployment(tx, tx.SndAddr)
 	case process.SCInvoking:
 		return txProc.processSCInvoking(tx, tx.SndAddr, tx.RcvAddr)
+	case process.BuiltInFunctionCall:
+		return txProc.processSCInvoking(tx, tx.SndAddr, tx.RcvAddr)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/metaProcess.go:122]: no need here. It is only SCInvoking - that is the only thing in metachain. This is needed for ESDTBurn which first makes a "-" in the sender shard as a built in function and calls the burn function from the metachain system SC.

File: process/transaction/shardProcess.go [hunk 3]
```diff
@@ -24,6 +24,8 @@ import (
 var log = logger.GetOrCreate("process/transaction")
 var _ process.TransactionProcessor = (*txProcessor)(nil)
 
+const minMetaTxExtraGasCost = uint64(1_000_000)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/shardProcess.go:27]: it is the minimum. as you check whether a transaction is good for metachain - it must have at least 1MIL gas.

File: process/transaction/metaProcess_test.go [hunk 4]
```diff
@@ -214,16 +158,9 @@ func TestMetaTxProcessor_ProcessCheckNotPassShouldErr(t *testing.T) {
...
-		adb,
-		createMockPubkeyConverter(),
-		mock.NewOneShardCoordinatorMock(),
-		&mock.SCProcessorMock{},
-		&mock.TxTypeHandlerMock{},
-		createFreeTxFeeHandler(),
-	)
+	args := createMockNewMetaTxArgs()
+	args.Accounts = adb
+	execTx, _ := txproc.NewMetaTxProcessor(args)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/metaProcess_test.go:163]: done

File: process/transaction/metaProcess_test.go [hunk 5]
```diff
@@ -252,21 +189,15 @@ func TestMetaTxProcessor_ProcessMoveBalancesShouldCallProcessIfError(t *testing.
...
+	args.ScProcessor = &mock.SCProcessorMock{
+		ProcessIfErrorCalled: func(acntSnd state.UserAccountHandler, txHash []byte, tx data.TransactionHandler, returnCode string, returnMessage []byte, snapshot int) error {
+			called = true
+			return nil
 		},
-		&mock.TxTypeHandlerMock{},
-		createFreeTxFeeHandler(),
-	)
+	}
+	execTx, _ := txproc.NewMetaTxProcessor(args)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/metaProcess_test.go:200]: done

File: process/transaction/metaProcess_test.go [hunk 6]
```diff
@@ -434,16 +357,12 @@ func TestMetaTxProcessor_ProcessTransactionScTxShouldNotBeCalledWhenAdrDstIsNotI
...
-		scProcessorMock,
-		computeType,
-		createFreeTxFeeHandler(),
-	)
+	args := createMockNewMetaTxArgs()
+	args.Accounts = adb
+	args.ScProcessor = scProcessorMock
+	args.TxTypeHandler = computeType
+	args.ShardCoordinator = shardCoordinator
+	execTx, _ := txproc.NewMetaTxProcessor(args)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/metaProcess_test.go:365]: done

File: process/transaction/metaProcess_test.go [hunk 7]
```diff
@@ -361,18 +287,15 @@ func TestMetaTxProcessor_ProcessTransactionScTxShouldReturnErrWhenExecutionFails
...
+	args.ScProcessor = scProcessorMock
+	args.TxTypeHandler = &mock.TxTypeHandlerMock{
+		ComputeTransactionTypeCalled: func(tx data.TransactionHandler) (transactionType process.TransactionType) {
 			return process.SCInvoking
-		}},
-		createFreeTxFeeHandler(),
-	)
+		},
+	}
+	execTx, _ := txproc.NewMetaTxProcessor(args)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/metaProcess_test.go:298]: done

File: process/transaction/metaProcess_test.go [hunk 8]
```diff
@@ -308,20 +239,15 @@ func TestMetaTxProcessor_ProcessTransactionScTxShouldWork(t *testing.T) {
...
+	args.Accounts = adb
+	args.ScProcessor = scProcessorMock
+	args.TxTypeHandler = &mock.TxTypeHandlerMock{
+		ComputeTransactionTypeCalled: func(tx data.TransactionHandler) (transactionType process.TransactionType) {
+			return process.SCInvoking
 		},
-		createFreeTxFeeHandler(),
-	)
+	}
+	execTx, _ := txproc.NewMetaTxProcessor(args)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/metaProcess_test.go:250]: done

File: vm/systemSmartContracts/esdt.go [hunk 9]
```diff
@@ -346,6 +341,12 @@ func (e *esdt) burn(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
 		return vmcommon.UserError
 	}
 
+	err = e.eei.UseGas(args.GasProvided)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:347]: done

File: vm/systemSmartContracts/esdt.go (hunk 1 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:410]: done

**Review by BeniaminDrasovean — APPROVED** [review]

File: cmd/node/factory/structs.go [hunk 10]
```diff
@@ -1864,21 +1865,34 @@ func newMetaBlockProcessor(
...
+	argsNewMetaTxProcessor := transaction.ArgsNewMetaTxProcessor{
+		Hasher:           core.Hasher,
+		Marshalizer:      core.InternalMarshalizer,
+		Accounts:         stateComponents.AccountsAdapter,
+		PubkeyConv:       stateComponents.AddressPubkeyConverter,
+		ShardCoordinator: shardCoordinator,
+		ScProcessor:      scProcessor,
+		TxTypeHandler:    txTypeHandler,
+		EconomicsFee:     economicsData,
+		ESDTEnableEpoch:  generalSettingsConfig.MetaProtectionEnableEpoch,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go:1877]: done

File: process/transaction/metaProcess.go [hunk 11]
```diff
@@ -141,6 +159,12 @@ func (txProc *metaTxProcessor) processSCInvoking(
 	return txProc.scProcessor.ExecuteSmartContractTransaction(tx, acntSrc, acntDst)
 }
 
+// EpochConfirmed is called whenever a new epoch is confirmed
+func (txProc *metaTxProcessor) EpochConfirmed(epoch uint32) {
+	txProc.flagESDTEnabled.Toggle(epoch >= txProc.esdtEnableEpoch)
+	log.Debug("txProcessor: relayed transactions", "enabled", txProc.flagESDTEnabled.IsSet())
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/metaProcess.go:165]: done

**Review by BeniaminDrasovean — DISMISSED** [review]

File: process/transaction/metaProcess.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/transaction/metaProcess.go]: ok

File: process/transaction/shardProcess.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/transaction/shardProcess.go:27]: But then you check in line 369 this way:
```
		if tx.GasLimit > txProc.economicsFee.ComputeGasLimit(tx)+minMetaTxExtraGasCost {
			return process.ErrInvalidMetaTransaction
		}

```
This way if gas exceeds computed gas limit + minMetaTxExtraGasCost then the transaction is considered invalid. So this should either be renamed max.. or the condition is not good here

[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/shardProcess.go:27]: condition is wrong here. good find. now will have to change some tests ==))

**Review by BeniaminDrasovean — APPROVED** [review]

File: core/constants.go [hunk 12]
```diff
@@ -548,3 +548,6 @@ const MultiplyFactorForScCall = 3
 // MaxRoundsWithoutCommittedStartInEpochBlock defines the maximum rounds to wait for start in epoch block to be committed,
 // before a special action to be applied
 const MaxRoundsWithoutCommittedStartInEpochBlock = 50
+
+// MinMetaTxExtraGasCost is the constant defines for minimum gas value to be sent in meta transaction
```
Comment (SebastianMarian) [inline core/constants.go:552]: "is the constant defined for" or "is the constant that defines the minimum..."

Comment (sasurobert) [reply core/constants.go:552]: defined for :) done.

**Review by SebastianMarian — APPROVED** [review]

File: vm/systemSmartContracts/esdt.go [hunk 13]
```diff
@@ -107,7 +107,7 @@ func (e *esdt) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
 		return e.issue(args)
 	case "issueProtected":
```
Comment (camilbancioiu) [inline vm/systemSmartContracts/esdt.go:108]: Optionally, these strings should also have constants, just like "burn".

**Review by SebastianMarian — DISMISSED** [review]

File: cmd/node/factory/structs.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go]: done

File: process/transaction/metaProcess.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/metaProcess.go]: done

**Review by BeniaminDrasovean — DISMISSED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

File: vm/systemSmartContracts/eei.go [hunk 14]
```diff
@@ -137,8 +138,29 @@ func (host *vmContext) GetBalance(addr []byte) *big.Int {
...
-func (host *vmContext) SendGlobalSettingToAll(_ []byte, _ []byte) {
-	//TODO: implement this
+func (host *vmContext) SendGlobalSettingToAll(_ []byte, input []byte) {
+	outputTransfer := vmcommon.OutputTransfer{
+		Value:    big.NewInt(0),
+		Data:     input,
+		CallType: vmcommon.DirectCall,
+	}
+
+	for i := uint8(0); i < uint8(host.blockChainHook.NumberOfShards()); i++ {
```
Comment (camilbancioiu) [inline vm/systemSmartContracts/eei.go:148]: Recommend to use `uint16` and take the last 2 bytes into account when creating these `outputAccounts`. The loop is currently limited to 256 shards (sufficient for the near future, but not future-proof, as third-party shards might be theoretically added in the future and exceed 256). Using 2 bytes is also consistent with the current sharding mask suffix.

Comment (sasurobert) [reply vm/systemSmartContracts/eei.go:148]: we have to change in a lot of places if we make it bigger

**Pull request merged**